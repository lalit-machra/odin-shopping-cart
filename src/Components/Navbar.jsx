import { Link } from 'react-router-dom';
import { PropTypes } from 'prop-types';

export default function Navbar(cart) {
  // if (typeof itemsInCart !== 'Array') {
  //   return (
  //     <div className="navbar">
  //       <Link to="/">Home</Link>
  //       <Link to="/shopPage">Shop</Link>
  //     </div>
  //   );
  // } else {
  return (
    <div className="navbar">
      <Link to="/">Home</Link>
      <Link to="/shopPage">Shop</Link>
      {cart.itemsInCart != undefined ? 
        <Link to="/shoppingCart" state={{cartItems: cart.itemsInCart}}>Cart ({cart.itemsInCart.length})</Link>
      : ''}
    </div>
  );
  // }
}

Navbar.propTypes = {
  cart: PropTypes.number.isRequired,
  req: PropTypes.string.isRequired,
}