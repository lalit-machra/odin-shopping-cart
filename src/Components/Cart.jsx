import { useLocation } from 'react-router-dom';
import Navbar from './Navbar';
import styles1 from '../styles/cart.module.css';

export default function Cart() {
  const location = useLocation();
  const itemsInCart = location.state.cartItems;
  console.log(itemsInCart)
  let totalAmount = itemsInCart.reduce((total, item) => {
    return (total += item.value);
  }, 0).toFixed(2);

  let itemsList = itemsInCart.map(item => 
    <tr key={item.name}>
      <td className={styles1.tableData}>{item.name}</td>
      <td className={styles1.tableData}>${item.price}</td>
      <td className={styles1.tableData}>{item.quantity}</td>
      <td className={styles1.tableData}>${item.value}</td>
    </tr>
  )

  if (totalAmount != 0) {
    return (
      <>
        <Navbar></Navbar>
        <h1>Summary</h1>
        <table className={styles1.cartDataTable}>
          <thead>
            <tr>
              <th className={styles1.tableHeader}>Item</th>
              <th className={styles1.tableHeader}>Price</th>
              <th className={styles1.tableHeader}>Quantity</th>
              <th className={styles1.tableHeader}>Total</th>
            </tr>
          </thead>
          <tbody>
            {itemsList}
            <tr>
              <td className={styles1.tableData}></td>
              <td className={styles1.tableData}></td>
              <td className={styles1.tableData}></td>
              <td className={styles1.totalAmount}>${totalAmount}</td>
            </tr>
          </tbody>
        </table>
        <button className={styles1.payBtn}>Pay ${totalAmount}</button>
      </>
    );
  } else {
    return (
      <>
        <Navbar></Navbar>
        <h1>Your cart is empty!!!</h1>
      </>
    );
  }
}