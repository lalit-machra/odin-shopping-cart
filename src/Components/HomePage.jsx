import Navbar from './Navbar';

export default function HomePage() {
  return (
    <>
      <Navbar></Navbar>
      <h1>Welcome to ABc, the one stop solution for all your shopping needs</h1>
      <img src="src/images/online-store.jpg" className="storeImg"></img>
      <h2>
        Choose among 400+ categories and 1,00,000+ items, all quality checked
        to give you a fulfilling experience
      </h2>
      <h2>Features we offer:</h2>
      <ul>
        <li>Wide variety of items</li>
        <li>Premium Quality</li>
        <li>10 days return policy</li>
        <li>Cash on delivery</li>
        <li>24 hours contact support</li>
      </ul>
    </>
  );
}