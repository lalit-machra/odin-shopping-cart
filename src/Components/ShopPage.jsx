import { default as Navbar } from './Navbar';
import { useState, useEffect } from 'react';

function ShopPage() {
  const [items, setItems] = useState([]);
  const [itemsInCart, setItemsInCart] = useState([]);
  useEffect(() => {
    async function getData() {
      const data = await (await fetch('https://fakestoreapi.com/products')).json();
      let dataArr = [];
      data.slice(0, 12).forEach((item) => {
        dataArr.push(item);
      });
      setItems(dataArr);
    }
    getData();
  }, []);
  
  let domData = items.map((item, index) => {
    return (
    <div className="item" key={index}>
      <img src={item.image} className="itemImg"></img>
      <h2>{item.title}</h2>
      <h3>${item.price}</h3>
      <label>Quantity: </label>
      <input type="number" className="itemCount" defaultValue="1" min="1"></input>
      <button className="addToCartBtn" onClick={(e) => hanldeItemsInCart(e.target.parentNode)}>Add to Cart</button>
    </div>);
  });

  function hanldeItemsInCart(itemDiv) {
    let itemName = itemDiv.querySelector("h2").innerText;
    let itemQuant = itemDiv.querySelector("input").value;
    let itemPrice = +itemDiv.querySelector("h3").innerText.slice(1);
    let itemValue = itemQuant * itemPrice;
    setItemsInCart(itemsInCart => [...itemsInCart, {name: itemName, quantity: itemQuant, price: itemPrice, value: itemValue}]);
  }

  return (
    <>
      <Navbar itemsInCart={itemsInCart}></Navbar>
      <h1>ABc Online Store</h1>
      <div className="items">
        {domData}
      </div>
    </>
  );
}

export default ShopPage;