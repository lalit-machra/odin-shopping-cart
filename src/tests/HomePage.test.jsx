import { render, screen } from '@testing-library/react';
import { describe, it, expect } from 'vitest';
import { MemoryRouter } from 'react-router-dom';
import { default as HomePage } from '../Components/HomePage';

describe("HomePage component", () => {
  it("renders homepage correctly", () => {
    const { container } = render(
      <MemoryRouter>
        <HomePage></HomePage>
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });

  it("renders title of the homepage", () => {
    render(
      <MemoryRouter>
        <HomePage></HomePage>
      </MemoryRouter>
    );
    expect(screen.getByRole("heading", {name: /Welcome to ABc/})).toBeInTheDocument();
  });

  it("does not display cart option in navbar", () => {
    render(
      <MemoryRouter>
        <HomePage></HomePage>
      </MemoryRouter>
    );
    expect(screen.queryByRole("link", {name: /Cart/})).not.toBeInTheDocument();
    expect(screen.getByRole("link", {name: /Home/})).toBeInTheDocument();
  });
});