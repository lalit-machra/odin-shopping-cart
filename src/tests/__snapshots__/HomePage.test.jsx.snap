// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HomePage component > renders homepage correctly 1`] = `
<div>
  <div
    class="navbar"
  >
    <a
      data-discover="true"
      href="/"
    >
      Home
    </a>
    <a
      data-discover="true"
      href="/shopPage"
    >
      Shop
    </a>
  </div>
  <h1>
    Welcome to ABc, the one stop solution for all your shopping needs
  </h1>
  <img
    class="storeImg"
    src="src/images/online-store.jpg"
  />
  <h2>
    Choose among 400+ categories and 1,00,000+ items, all quality checked to give you a fulfilling experience
  </h2>
  <h2>
    Features we offer:
  </h2>
  <ul>
    <li>
      Wide variety of items
    </li>
    <li>
      Premium Quality
    </li>
    <li>
      10 days return policy
    </li>
    <li>
      Cash on delivery
    </li>
    <li>
      24 hours contact support
    </li>
  </ul>
</div>
`;
